iroha_actor/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+    }
+}
+
+impl<A: Actor> Addr<A> {
+    fn new(sender: mpsc::Sender<Envelope<A>>) -> Self {
+        Self {
+            sender,
+            #[cfg(feature = "deadlock_detection")]
+            actor_id: ActorId::from(&task::current()),
+        }
+    }
+
+    /// Send a message and wait for an answer.
+    /// # Errors
+    /// Fails if noone will send message
+    pub async fn send<M>(&self, message: M) -> Result<M::Result, RecvError>
+    where
+        M: Message + Send + 'static,
+        M::Result: Send,
+        A: Handler<M>,
+    {
+        let (sender, reciever) = oneshot::channel(1);
+        let envelope = SyncEnvelopeProxy::pack(message, Some(sender));
+        #[cfg(feature = "deadlock_detection")]
+        deadlock::r#in(self.actor_id).await;
+        self.sender.send(envelope).await;
+        let result = reciever.recv().await;
+        #[cfg(feature = "deadlock_detection")]
+        deadlock::out(self.actor_id).await;
+        result
+    }
+
+    /// Send a message and wait for an answer.
+    /// # Errors
+    /// Fails if queue is full or actor is disconnected
+    #[allow(clippy::result_unit_err)]
+    pub async fn do_send<M>(&self, message: M)
+    where
+        M: Message + Send + 'static,
+        M::Result: Send,
+        A: Handler<M>,
+    {
+        let envelope = SyncEnvelopeProxy::pack(message, None);
+        self.sender.send(envelope).await
+    }
+
+    /// Constructs recipient for sending only specific messages (without answers)
+    pub fn recipient<M>(&self) -> Recipient<M>
+    where
+        M: Message<Result = ()> + Send + 'static,
+        A: Handler<M>,
+    {
+        Recipient(Box::new(self.clone()))
+    }
+}
+
+#[allow(missing_debug_implementations)]
+/// Address of actor. Can be used to send messages to it.
+pub struct Recipient<M: Message<Result = ()>>(Box<dyn Sender<M> + Sync + Send + 'static>);
+
+impl<M: Message<Result = ()> + Send> Recipient<M> {
+    /// Send message to actor
+    pub async fn send(&self, m: M) {
+        self.0.send(m).await
+    }
+}
+
+#[async_trait::async_trait]
+trait Sender<M: Message<Result = ()>> {
+    async fn send(&self, m: M);
+}
+
+#[async_trait::async_trait]
+impl<A, M> Sender<M> for Addr<A>
+where
+    M: Message<Result = ()> + Send + 'static,
+    M::Result: Send,
+    A: Handler<M>,
+{
+    async fn send(&self, m: M) {
+        self.do_send(m).await
+    }
+}
+
+/// Actor trait
+#[async_trait::async_trait]
+pub trait Actor: Send + Sized + 'static {
+    /// Capacity of actor queue
+    const CAPACITY: usize = 100;
+
+    /// At start hook of actor
+    async fn started(&mut self, _ctx: &mut Context<Self>) {}
+
+    /// At stop hook of actor
+    async fn stopped(&mut self, _ctx: &mut Context<Self>) {}
+
+    /// Start actor
+    fn start(mut self) -> Addr<Self> {
+        let (sender, reciever) = mpsc::channel(Self::CAPACITY);
+        let addr = Addr::new(sender);
+        let mut ctx = Context::new(addr.clone());
+        #[allow(clippy::expect_used)]
```

Comment:
I don't think it is used here...

Reply:
Yeah. Forgot to remove it. Tried to use task builder for constructing task with specific names. It is actually infallible (looked at the source code, and they never fail). Should i add it here?

Reply:
I guess no need as we are migrating to tokio and there it might be different